Fix logic in zone09

tgreenx opened the pull request:
## Purpose

This PR updates the implementation of Zone09 due to an oversight in https://github.com/zonemaster/zonemaster-engine/pull/1109. The implementation was not behaving exactly as per the specification and could output unintended messages tags.

## Context

Issue found during release testing (thanks @blacksponge)

## Changes

- Looping through all RRsets removed since it is only needed to inspect one RRset
- Add control variable to check if there is a NULL MX in the RRset

## How to test this PR

Unit tests should pass.

matsduf commented:
This has been tested on the environment described in #1113.

**First issue:**
```
$ zonemaster-cli INCONSISTENT-MX-DATA.zone09.xa --raw  --test Zone/zone09 --hints COMMON/hintfile
   0.10 WARNING   Z09_INCONSISTENT_MX_DATA   
```
Z09_MX_DATA is also expected.

```
$ dig @127.3.0.53 INCONSISTENT-MX-DATA.zone09.xa ns +short
ns2.INCONSISTENT-MX-DATA.zone09.xa.
ns1.INCONSISTENT-MX-DATA.zone09.xa.

$ dig @127.3.0.53 ns1.INCONSISTENT-MX-DATA.zone09.xa +short
127.19.9.31

$ dig @127.3.0.53 ns2.INCONSISTENT-MX-DATA.zone09.xa +short
127.19.9.32

$ dig @127.19.9.31 INCONSISTENT-MX-DATA.zone09.xa mx +short
10 mail.xa.
10 liam.xa.

$ dig @127.19.9.32 INCONSISTENT-MX-DATA.zone09.xa mx +short
10 liam.xa.
```
**Second issue:**
```
$ zonemaster-cli INCONSISTENT-MX.zone09.xa --raw  --test Zone/zone09 --hints COMMON/hintfile
   0.09 WARNING   Z09_INCONSISTENT_MX   
   0.09 NOTICE    Z09_MISSING_MAIL_TARGET   
```
Z09_MISSING_MAIL_TARGET is unexpected. Z09_MX_FOUND and Z09_NO_MX_FOUND are expected.

```
$ dig @127.3.0.53 INCONSISTENT-MX.zone09.xa ns +short
ns2.INCONSISTENT-MX.zone09.xa.
ns1.INCONSISTENT-MX.zone09.xa.

$ dig @127.3.0.53 ns1.INCONSISTENT-MX.zone09.xa +short
127.19.9.31

$ dig @127.3.0.53 ns2.INCONSISTENT-MX.zone09.xa +short
127.19.9.32

$ dig @127.19.9.31 INCONSISTENT-MX.zone09.xa mx +short
10 mail.xa.

$ dig @127.19.9.32 INCONSISTENT-MX.zone09.xa mx +short
(NODATA)
```

matsduf reviewed (changes requested):
On line 1003 (https://github.com/zonemaster/zonemaster-engine/pull/1163/files#diff-5acc06bb09a978de03ad11cb90fc14f6c0edc9497d87caf4a14bc434f2a6f352R1003) is says
```
        unless ( $zone->name eq '.' or $zone->name->next_higher() eq '.' or $zone->name =~ /\.arpa/ ){
```
It should be `$zone->name =~ /\.arpa$/ `

matsduf reviewed (commented):
On line 1002 (https://github.com/zonemaster/zonemaster-engine/pull/1163/files#diff-5acc06bb09a978de03ad11cb90fc14f6c0edc9497d87caf4a14bc434f2a6f352R1002) it says
```
    if ( scalar @no_mx_set ){
```
It should rather be
```
    if ( scalar @no_mx_set  and !scalar %mx_set ){
```

tgreenx commented:
> On line 1002 (https://github.com/zonemaster/zonemaster-engine/pull/1163/files#diff-5acc06bb09a978de03ad11cb90fc14f6c0edc9497d87caf4a14bc434f2a6f352R1002) it says
> 
> ```
>     if ( scalar @no_mx_set ){
> ```
> 
> It should rather be
> 
> ```
>     if ( scalar @no_mx_set  and !scalar %mx_set ){
> ```

Fixed (I went with an "elsif" instead). This was (partly) causing issue number 2 that you described.

> On line 1003 (https://github.com/zonemaster/zonemaster-engine/pull/1163/files#diff-5acc06bb09a978de03ad11cb90fc14f6c0edc9497d87caf4a14bc434f2a6f352R1003) is says
> 
> ```
>         unless ( $zone->name eq '.' or $zone->name->next_higher() eq '.' or $zone->name =~ /\.arpa/ ){
> ```
> 
> It should be `$zone->name =~ /\.arpa$/ `

Updated.

> [ ... ]
> 
> ```
> $ zonemaster-cli INCONSISTENT-MX-DATA.zone09.xa --raw  --test Zone/zone09 --hints COMMON/hintfile
>    0.10 WARNING   Z09_INCONSISTENT_MX_DATA   
> ```
> 
> Z09_MX_DATA is also expected.
>
> [ ... ] 

In this case I'm not sure why it is behaving this way. The code that outputs data for this message tag is directly after the output of that message tag (see line xxx).

We have unitary tests `t/Test-zone09-B.t` and  `t/Test-zone09-C.t` for that kind of message tag right now and it works well there.

```
$ zonemaster-cli --raw --show-testcase --test zone/zone09 --level INFO --ns zonemaster01-prd.iis.se --ns zonemaster05-prd.iis.se xb
   0.00 INFO      UNSPECIFIED    GLOBAL_VERSION   version=v4.5.1
   0.57 WARNING   ZONE09         Z09_INCONSISTENT_MX_DATA
   0.57 INFO      ZONE09         Z09_MX_DATA   mailtarget_list=mx2.iis.se.;mx1.iis.se.; ns_ip_list=45.155.96.81
   0.57 INFO      ZONE09         Z09_MX_DATA   mailtarget_list=mahimahi.ripe.net.;molamola.ripe.net.; ns_ip_list=45.155.96.85

$ dig @45.155.96.81 MX xb +short
5 mx2.iis.se.
5 mx1.iis.se.

$ dig @45.155.96.85 MX xb +short
5 molamola.ripe.net.
0 mahimahi.ripe.net.
```

```
$ zonemaster-cli --raw --show-testcase --test zone/zone09 --level INFO --ns zonemaster01-prd.iis.se --ns zonemaster05-prd.iis.se xc
   0.00 INFO      UNSPECIFIED    GLOBAL_VERSION   version=v4.5.1
   0.59 WARNING   ZONE09         Z09_INCONSISTENT_MX_DATA
   0.59 INFO      ZONE09         Z09_MX_DATA   mailtarget_list=mx1.iis.se.;mx2.iis.se.; ns_ip_list=45.155.96.81
   0.59 INFO      ZONE09         Z09_MX_DATA   mailtarget_list=molamola.ripe.net.;.; ns_ip_list=45.155.96.85

$ dig @45.155.96.81 MX xc +short
5 mx1.iis.se.
5 mx2.iis.se.

$ dig @45.155.96.85 MX xc +short
5 molamola.ripe.net.
0 .
```

> 
> **Second issue:**
> 
> ```
> $ zonemaster-cli INCONSISTENT-MX.zone09.xa --raw  --test Zone/zone09 --hints COMMON/hintfile
>    0.09 WARNING   Z09_INCONSISTENT_MX   
>    0.09 NOTICE    Z09_MISSING_MAIL_TARGET   
> ```
> 
> Z09_MISSING_MAIL_TARGET is unexpected. Z09_MX_FOUND and Z09_NO_MX_FOUND are expected.
> 

Z09_MISSING_MAIL_TARGET should be fixed now (this was due to the missing `elsif` mentioned above). For the other half I'm not sure either. We have unitary test `t/Test-zone09-A.t` for those messages tags and it works well there.

> [ ... ]
> 
> **Third issue:**
> 
> ```
> $ zonemaster-cli mx-data.zone09.xa --raw  --test Zone/zone09 --hints COMMON/hintfile
> (empty)
> ```
> 
> Z09_MX_DATA is expected.
> 
> [ ... ]

Not sure either. Unitary test `t/Test-zone09-F.t` should cover this case and it works well there too.

------------------------------

Awaiting further debugging to investigate the issues.

matsduf commented:
Issue 1:
```
zonemaster-cli INCONSISTENT-MX-DATA.zone09.xa --raw  --test Zone/zone09 --hints COMMON/hintfile

$VAR1 = 'ns1.inconsistent-mx-data.zone09.xa';
$VAR2 = [
          '127.19.9.31',
          'fda1:b2:c3:0:127:19:9:31'
        ];
$VAR3 = 'ns2.inconsistent-mx-data.zone09.xa';
$VAR4 = [
          '127.19.9.32',
          'fda1:b2:c3:0:127:19:9:32'
        ];

$VAR1 = '127.19.9.31';
$VAR2 = [
          bless( do{\(my $o = '93962842591344')}, 'Zonemaster::LDNS::RR::MX' ),
          bless( do{\(my $o = '93962843194432')}, 'Zonemaster::LDNS::RR::MX' )
        ];
$VAR3 = '127.19.9.32';
$VAR4 = [
          bless( do{\(my $o = '93962843655824')}, 'Zonemaster::LDNS::RR::MX' )
        ];
$VAR5 = 'fda1:b2:c3:0:127:19:9:32';
$VAR6 = [
          bless( do{\(my $o = '93962843853248')}, 'Zonemaster::LDNS::RR::MX' )
        ];
$VAR7 = 'fda1:b2:c3:0:127:19:9:31';
$VAR8 = [
          bless( do{\(my $o = '93962842823616')}, 'Zonemaster::LDNS::RR::MX' ),
          bless( do{\(my $o = '93962843457808')}, 'Zonemaster::LDNS::RR::MX' )
        ];
   0.13 CRITICAL  MODULE_ERROR   module=Zone; msg=Died at /usr/local/share/perl/5.34.0/Zonemaster/Engine/Test/Zone.pm line 915.
```

matsduf commented:
Issue 2:
```
$ zonemaster-cli INCONSISTENT-MX.zone09.xa --raw  --test Zone/zone09 --hints COMMON/hintfile
no_mx_set: 127.19.9.32;fda1:b2:c3:0:127:19:9:32
$VAR1 = 'ns1.inconsistent-mx.zone09.xa';
$VAR2 = [
          '127.19.9.31',
          'fda1:b2:c3:0:127:19:9:31'
        ];
$VAR3 = 'ns2.inconsistent-mx.zone09.xa';
$VAR4 = [
          '127.19.9.32',
          'fda1:b2:c3:0:127:19:9:32'
        ];

$VAR1 = 'fda1:b2:c3:0:127:19:9:31';
$VAR2 = [
          bless( do{\(my $o = '93955562515680')}, 'Zonemaster::LDNS::RR::MX' )
        ];
$VAR3 = '127.19.9.31';
$VAR4 = [
          bless( do{\(my $o = '93955562252768')}, 'Zonemaster::LDNS::RR::MX' )
        ];
   0.12 CRITICAL  MODULE_ERROR   module=Zone; msg=Died at /usr/local/share/perl/5.34.0/Zonemaster/Engine/Test/Zone.pm line 915.
```

matsduf commented:
My debugging was run using 81e0fedf13d59d103c4f782cd0ecbabc478d1de5, not the update today.

tgreenx commented:
> Issue 1:
> 
> [ ... ]
> Issue 2:
>
> [ ... ]
> Issue 3:
>
> [ ... ]

You are missing `--level INFO` in your command, that is why the messages are not showing. This is the default log level for these messages, see https://github.com/zonemaster/zonemaster-engine/blob/develop/share/profile.json#L479-L482

matsduf commented:
> I suspect all issues are from the same problem: you are missing `--level INFO` in your command, that is why the messages are not showing.

Yes, we missed that. I have checked the updated code (with the corrections in this PR) and now all tests pass, except for one that cannot yet be configured correctly (error in test zone, not in test case implementation).
